File: hilbertutils.py
def xy2d(n: int, x: int, y: int) -> int:
    """Takes a 2d point and returns the point's location on the hilbert curve

    Args:
        n (int): The dimension of the hilbert curve
        x (int): The x coordinate of the 2d point
        y (int): The y coordinate of the 2d point

    Returns:
        int: the point on the hilbert curve
    """
    rx: int = 0
    ry: int = 0
    s: int = int(n / 2)
    d: int = 0

    while s > 0:
        rx = int((x & s) > 0)
        ry = int((y & s) > 0)

        d += s*s*((3 * rx) ^ ry)
        x,y = rot(n, x, y, rx, ry)
        s = s // 2

    return d

def d2xy(n: int, d: int) -> tuple([int, int]):
    """ Takes a point on the n-order hilbert curve and returns it's coordinates in 2d

    Args:
        n (int): The dimension of the hilbert curve
        d (int): The point on the hilbert curve

    Returns:
        tuple([int, int]): x,y coordinates 
    """
    rx: int = 0
    ry: int = 0
    s: int = 1
    t: int = d

    x: int = 0
    y: int = 0

    while s < n:
        rx = 1 & (t // 2)
        ry = 1 & (t ^ rx)

        x,y = rot(s, x, y, rx, ry)

        x += s * rx
        y += s * ry

        t = t // 4
        s *= 2
    
    return x,y

def rot(n : int, x: int, y: int, rx: int, ry: int) -> tuple([int, int]):
    """Given a 2D point, (x,y), return the rotated version of the point

    Args:
        n (int): Order of the hilbert curve

    Returns:
        tuple([int, int]): (x,y) rotated version of the 2D point
    """
    if ry == 0:
        if rx == 1:
            x = n - 1 - x
            y = n - 1 - y
        
        x,y = y,x
        
    
    return x,y


def generate_hilbert_path(n: int) -> list:
    """Generates a hilbert path and writes it to a file if it doesn't yet exist.
    If it does, then load the file as this function can take a long time to run.

    Args:
        n (int): Order of the hilbert curve.
    """

    # TODO: Check for existence of cached hilbert path

    path = [0] * n * n
    for d in range(len(path)):
        path[d] = d2xy(n, d)
    return path

if __name__ == '__main__':

    print("Running Tests")
    n_lst = [int(2**i) for i in range(4,10)]
    for n in n_lst:
        for x in range(n):
            for y in range(n):
                d = xy2d(n, x, y)
                #print(f'x: {x:3d} | y: {y:3d} | d: {d:3d}')

                xd,yd = d2xy(n - 1, xy2d(n, x, y))

                assert xd == x, f"xy2d or d2xy error: x should be {x:3d} but got {xd:3d} instead"
                assert yd == y, f"xy2d or d2xy error: y should be {y:3d} but got {yd:3d} instead"
    print("Tests passed")

File: scripts/hilbertutils.js
/** Reimplementation of the python algorithm */

/** 
 * Takes a 2d point and returns the point's location on the hilbert curve
 * @param   {int} n The dimension of the hilbert curve
 * @param   {int} x The x coordinate of the 2d point
 * @param   {int} y The y coordinate of the 2d point
 * @returns {int} d The point on the hilbert curve
 */
function xy2d(n, x, y) {
    var rx = 0;
    var ry = 0;
    var s = Math.floor(n / 2);
    var d = 0;

    while (s > 0) {
        rx = (x & s) > 0;
        ry = (y & s) > 0;

        d += s*s*((3*rx) ^ ry);
        [x,y] = rot(s, x, y, rx, ry);

        s = Math.floor(s / 2);
    }

    return d;
}

/**
 * Takes a point on the n-order hilbert curve and returns it's coordinates in 2d
 * @param {int} n The dimension of the hilbert curve
 * @param {int} d The point on the hilbert curve
 * @returns {[int, int]} x,y the 2d point on the hilbert curve
 */
function d2xy(n, d) {
    var rx = 0;
    var ry = 0;
    var s = 1;
    var t = d;

    var x = 0;
    var y = 0;

    while (s < n) {
        rx = 1 & (Math.floor(t / 2));
        ry = 1 & (t ^ rx);

        [x,y] = rot(s, x, y, rx, ry);

        x += s * rx;
        y += s * ry;

        t = Math.floor(t / 4);
        s = s * 2;
    }

    return [x,y];
}

/**
 * 
 * @param   {int} n The dimension of the hilbert curve
 * @param   {int} x The x coordinate of the 2d point
 * @param   {int} y The y coordinate of the 2d point
 * @param   {int} rx rotate on x 
 * @param   {int} ry rotate on y
 * @returns {[int, int]} the rotated point
 */
function rot(n, x, y, rx, ry) {
    var t;
    if (ry == 0) {
        if (rx == 1) {
            x = n - 1 - x;
            y = n - 1 - y;
        }
        return [y, x];
    }

    return [x, y];
}

/**
 * Generates a d3 path adjusted according to size of canvas
 * @param   {int} n The dimension of the hilbert curve
 * @param   {int} size The dimensions of the canvas, typically the shortest side if non-square
 * @returns {d3.path} The translated 2d points of the hilbert curve
 */
function generatePath(n, size) {
    /**
     * param order (int): The order of the hilbert curve
     * param size (int): The actual dimensions of the canvas
     */
    var path = d3.path();

    var unit = size / (2**n);
    var x,y;

    if ((n % 2) == 1) {
        path.moveTo(0, unit/2);
    } else {
        path.moveTo(unit/2, 0);
    }
    

    path.lineTo(unit/2, unit/2);

    for (let index = 0; index < 2**n*2**n; index++) {
        [x,y] = d2xy(2**n, index);
        path.lineTo(x * unit +unit/2, y * unit + unit/2);
    }

    if ((n % 2) == 1) {
        path.lineTo(size, unit/2);
    } else {
        path.lineTo(size - unit/2, 0);
    }
    
    return path;
}

